### PR TITLE
feat: add initial requestThingDescription implementation

### PR DESCRIPTION
**Pull request by JKRhb**
This PR adds an initial implementation for the `requestThingDescription` method.

The implementation details probably still need some additional discussion – looking forward to your feedback here :) For now, I extended the `ProtocolClient` interface with a new `requestThingDescription` method that is expected to return a `Content` object that is deserialized and (supposed to be ) validated by the `WoT` implementation.

I think this latter part also touches #1055 – in my opinion, there should be some validation happening so that library users can be sure that they always receive a valid Thing Description. Maybe the algorithm(s) in the Scripting API would need to be adjusted then as well.

For now, I only implemented the protocol-specific `requestThingDescription` method for HTTP and CoAP, if the general design works for you more protocols could be added in follow-up PRs.

### CODE REVIEW

File: packages/binding-mqtt/src/mqtt-client.ts
Comment (egekorkan) [inline packages/binding-mqtt/src/mqtt-client.ts:168]: Technically, this can be implemented since an MQTT Thing exposed by node-wot publishes its TD with retain. See https://github.com/eclipse-thingweb/node-wot/blob/master/packages/binding-mqtt/src/mqtt-broker-server.ts#L141 It is not standardized afaik but I do not know what speaks against doing a subscription, getting a TD and unsubscribing. Of course it can be done in another PR, I leave it up to you @JKRhb

Comment (JKRhb) [reply packages/binding-mqtt/src/mqtt-client.ts:168]: Oh, yeah, very good point, I implemented something similar in `dart_wot` so I can confirm that this is working :) Maybe I would open a follow-up PR for this feature, though, since it might require some discussion regarding the URL-to-topic mapping 🤔

Comment (egekorkan) [reply packages/binding-mqtt/src/mqtt-client.ts:168]: Yup sure. Interestingly, this is also mandated by sparkplug with the birth messages (same as expose here). More reading is possible at https://sparkplug.eclipse.org/specification/version/3.0/documents/sparkplug-specification-3.0.0.pdf at page 21. Also, the newer version of it allows any payload to be published there (wink wink TD). 

In the end, as long as the developer provides the URL to fetch from, I think it is fine and no need for big discussions.

File: packages/core/src/wot-impl.ts
Comment (relu91) [inline packages/core/src/wot-impl.ts:49]: Unrelated comment. I found it strange that the function ContentToValue does not accept directly the `result` variable which is type Content. Something to record in an issue for future improvements.

File: packages/binding-file/src/file-client.ts
Comment (relu91) [inline packages/binding-file/src/file-client.ts:79]: Is there a blocker to implement this or is it just a time constraint? I think we can jus read the file pointed by the file URI.

**Review by relu91 — COMMENTED** [review]
I would fix Ege's point and I left a couple of comments below. Moreover, I wondering if we can somehow track the different TODOs comments before merging. Maybe when we reched two accepts we can create the issues and do a quick update of the comments with a link to the issues? Finally, maybe I missed why there is so much discrepancy between `coaps` and `coap` implementation?

File: packages/core/src/wot-impl.ts
Comment (JKRhb) [reply packages/core/src/wot-impl.ts:49]: Yeah, that is indeed a bit strange, apparently `contentToValue` expects input that implements the `ReadContent` interface which is not compatible with the regular `Content` class 🤔 It would definitely be better to have one unified `Content` API that is used across the codebase. Opening an issue for that sounds like a good idea to me!

File: packages/binding-mqtt/src/mqtt-client.ts
Comment (JKRhb) [reply packages/binding-mqtt/src/mqtt-client.ts:168]: Hmm, having another look at the MQTT client implementation after @relu91's comment I noticed that it actually does not even implement the `readResource` method where I think a similar logic would be used. So my proposal would be to implement both methods at the same time to arrive at a uniform implementation and keep the discussion focused – would that be okay for you, @relu91? Otherwise, adding the implementation here would also not be a problem, though :)

File: packages/binding-file/src/file-client.ts
Comment (JKRhb) [reply packages/binding-file/src/file-client.ts]: I added an implementation now in 3a7c741e48a155b01039b3745db65fedd402755f :) I think there might actually be some more room for simplifying the `FileClient` implementation, though, since the file extension determination is not actually needed when we assume a default Content-Type of `application/json` for `Form`s. However, this would constitute a breaking change.

**Comment by JKRhb** [comment]
> Finally, maybe I missed why there is so much discrepancy between `coaps` and `coap` implementation?

That is actually because we are currently using two different libraries for CoAP (`node-coap`) and CoAPS (`node-coap-client`) since `node-coap` does not support DTLS :/ A mid-term goal for me personally is definitely to arrive at a unified implementation that covers not only CoAP and CoAPS but also CoAP over TCP/TLS and WebSockets. For the time being, though, I think we need to live with this doubled CoAP implementation.

File: packages/core/src/wot-impl.ts
Comment (JKRhb) [inline packages/core/src/wot-impl.ts:54]: I got rid of the TODO regarding validation now by passing the `TDSchema` to the `contentToValue` method alongside the TD content. I hope that makes sense – more validation could probably be added later.

File: packages/binding-file/src/file-client.ts
Comment (danielpeintner) [inline packages/binding-file/src/file-client.ts:67]: Would be failing if `href` split returns array with zero size, right? Are we fine with it?

**Review by danielpeintner — APPROVED** [review]
I think for an initial implementation it is fine as is.

I added some comments below

File: packages/binding-file/src/file-client.ts
Comment (danielpeintner) [inline packages/binding-file/src/file-client.ts:31]: the comment is somehow broken

Comment (JKRhb) [reply packages/binding-file/src/file-client.ts]: Thank you :) Should be fixed by 907b429a5e4400de62a5999e59a24c18850e62a4

File: packages/binding-file/src/file-client.ts
Comment (JKRhb) [reply packages/binding-file/src/file-client.ts:67]: At the moment, this behavior is consistent with the previous version, I just refactored it a bit (previously, `filepath` was an array that got accessed this way twice). Adding a check or a fallback value sounds good to me, though – should this be done in this PR? Or should we follow up with another one to do so?

File: packages/core/src/wot-impl.ts
Comment (JKRhb) [reply packages/core/src/wot-impl.ts]: Good point, I now replaced this line with the `extractScheme` helper method in b67162da2984f44c487054258d9c74736b5779e8, which makes the implementation cleaner and more consistent with the rest of the codebase.

File: packages/binding-file/src/file-client.ts
Comment (danielpeintner) [reply packages/binding-file/src/file-client.ts:67]: I "think" we can also do it in this PR

Comment (JKRhb) [reply packages/binding-file/src/file-client.ts]: I now replaced the splitting in this line with accessing a parsed `URL` object. In this context, I noticed that the previous implementation might not have been completely correct since it used the *host* component of the parsed URL instead of the *path* component (which, at least according to [Wikipedia](https://en.wikipedia.org/wiki/File_URI_scheme) is "never correct"). It seems like the README of the file binding also needs to be adjusted in this regard.

File: packages/binding-file/src/file-client.ts
Comment (relu91) [reply packages/binding-file/src/file-client.ts]: It is good! I would like to keep that auto detection of content type... TD 2.0 might support other formats than the default.

File: packages/core/src/wot-impl.ts
Comment (relu91) [reply packages/core/src/wot-impl.ts:54]: As I recall `contentToValue` method does not do the validation. It uses the data schema only as a hint when converting. The proper way to handle this should use AJV here. But we have to think this correctly, given #1062 .

File: packages/binding-file/src/file-client.ts
Comment (relu91) [reply packages/binding-file/src/file-client.ts]: I remember that parsing File URIs is problematic. A couple of unit tests might be good. Can you add them?

Comment (JKRhb) [reply packages/binding-file/src/file-client.ts]: Sure, will do :)

File: packages/core/src/wot-impl.ts
Comment (JKRhb) [reply packages/core/src/wot-impl.ts]: In 5aeb45dde8507cb205e5f5c406e0b1f7528a15f1, I now incorporated the validation code used by @danielpeintner in #1169 instead

File: packages/binding-file/src/file-client.ts
Comment (JKRhb) [reply packages/binding-file/src/file-client.ts]: I now included the file client code in a separate PR: #1175 Once we have merged that one, I would perform a rebase to keep a clean and reviewable diff in this PR :)

Comment (danielpeintner) [reply packages/binding-file/src/file-client.ts]: Makes sense 👍

**Review by danielpeintner — APPROVED** [review]
LGTM

**Review by relu91 — APPROVED** [review]
Good enough for now. Let's improve it with future iterations!

**Pull request merged**
